### PR TITLE
subset lb: avoid partitioning host lists on worker threads

### LINKED ISSUE #6301: Race in subset LB host updates
During host updates on the worker threads, the subset LB calls `health()` on the list of all hosts to partition to the hosts into healthy/degraded/unhealthy vectors. Since `health()` reads from an atomic
bit mask that may be modified by another thread (ie by the health checker), this may produce an inconsistent view of a given HostSet, since `health()` is called multiple times for each host. 

I can think of a few ways of solving this:
1) Ensure we only call `health()` exactly once per host per subset update. This will ensure that the health values are consistent *within a given HostSet*. It would not guarantee that all subsets are updated with the same health values.
2) Infer the health values of hosts by determining which of the healthy/degraded/all hosts on the original
priority set they're in. This would ensure consistency between all subsets.

Leaning towards 2), so I'll see if I can get a PR ready.

### PR DESCRIPTION
**Pull request by snowp**
This changes the subset LB to not read host.health() on the worker
thread, instead relying on the partitioning made on the main thread.
This ensures consistency as host.health() might change due to health
checks, config updates, etc.

Also ensures consistent handling of metadata reads by caching the result
of the predicate. This is likely to come with a perf improvement, as
we're now making fewer metadata reads.

Fixes #6301

Signed-off-by: Snow Pettersen <snowp@squareup.com>

Risk Level: Medium, changes the algorithm for processing subset host updates.
Testing: Existing unit test coverage, there should be no change in behavior.
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by snowp** [comment]
Didn't add any tests since there were no behavior changes, but it might make sense to add a regression test that verifies that `health()` is not called during the update

**Comment by snowp** [comment]
The regression I was thinking about is the fact that calling `health()` multiple times for the same host. Since the value might change between calls, it would give rise to subtle bugs. Not sure if it's worth a test, wdyt? Maybe a comment is good enough?

**Comment by snowp** [comment]
/retest 

(timeout in mysql test)

**Comment by zuercher** [comment]
I see. Seems worth it if it's not too complicated to write. You might be able to piggy back on an existing subset lb test where the same host appears in multiple sets and just change a mock to error on multiple calls.

File: source/common/upstream/subset_lb.cc [hunk 1]
```diff
@@ -509,52 +509,43 @@ SubsetLoadBalancer::PrioritySubsetImpl::PrioritySubsetImpl(const SubsetLoadBalan
 void SubsetLoadBalancer::HostSubsetImpl::update(const HostVector& hosts_added,
                                                 const HostVector& hosts_removed,
                                                 std::function<bool(const Host&)> predicate) {
-  std::unordered_set<HostSharedPtr> predicate_added;
+  HostVectorSharedPtr hosts(new HostVector());
+  HostVectorSharedPtr healthy_hosts(new HostVector());
+  HostVectorSharedPtr degraded_hosts(new HostVector());
```
Comment (rgs1) [inline source/common/upstream/subset_lb.cc]: nit: isn't `std::make_shared()` preferred here?

File: source/common/upstream/subset_lb.cc [hunk 2]
```diff
@@ -509,52 +509,43 @@ SubsetLoadBalancer::PrioritySubsetImpl::PrioritySubsetImpl(const SubsetLoadBalan
...
+  // We use an unordered_set because this can potentially be in the tens of thousands.
+  std::unordered_set<const Host*> matching_hosts;
+
+  auto cached_predicate = [&matching_hosts](const auto& host) {
+    return matching_hosts.count(&host) == 1;
+  };
+
+  // TODO(snowp): If we had a unhealthyHosts() function we could avoid potentially traversing
+  // the list of hosts twice.
+  for (const auto host : original_host_set_.hosts()) {
```
Comment (rgs1) [inline source/common/upstream/subset_lb.cc]: isn't `const auto& host` cheaper here? [avoid copying the `shared_ptr`, which implies increasing refcount, etc]

File: source/common/upstream/subset_lb.cc [hunk 3]
```diff
@@ -509,52 +509,43 @@ SubsetLoadBalancer::PrioritySubsetImpl::PrioritySubsetImpl(const SubsetLoadBalan
...
+      matching_hosts.insert(host.get());
+      hosts->emplace_back(host);
     }
   }
 
-  HostVector filtered_removed;
-  for (const auto host : hosts_removed) {
-    if (predicate(*host)) {
-      filtered_removed.emplace_back(host);
+  for (const auto host : original_host_set_.healthyHosts()) {
```
Comment (rgs1) [inline source/common/upstream/subset_lb.cc]: ditto?

**Comment by snowp** [comment]
@zuercher I added the regression test finally, PTAL

**Pull request merged**